modules/mono/glue/GodotSharp/GodotSharp/Core/Projection.cs

```diff
@@ -0,0 +1,822 @@
+        {
+            Vector2 vpHe = GetViewportHalfExtends();
+            return vpHe.x / vpHe.y;
+        }
+
+        public Vector2 GetFarPlaneHalfExtends()
+        {
+            var res = GetProjectionPlane(Planes.Far).Intersect3(GetProjectionPlane(Planes.Right), GetProjectionPlane(Planes.Top));
+            return new Vector2(res.Value.x, res.Value.y);
+        }
+
+        public real_t GetFov()
+        {
+            Plane rightPlane = new Plane(x.w - x.x, y.w - y.x, z.w - z.x, -w.w + w.x).Normalized();
+            if (z.x == 0 && z.y == 0)
+            {
+                return Mathf.Rad2Deg(Mathf.Acos(Mathf.Abs(rightPlane.Normal.x))) * (real_t)2.0;
+            }
+            else
+            {
+                Plane leftPlane = new Plane(x.w + x.x, y.w + y.x, z.w + z.x, w.w + w.x).Normalized();
+                return Mathf.Rad2Deg(Mathf.Acos(Mathf.Abs(leftPlane.Normal.x))) + Mathf.Rad2Deg(Mathf.Acos(Mathf.Abs(rightPlane.Normal.x)));
+            }
+        }
+
+        public static real_t GetFovy(real_t fovx, real_t aspect)
+        {
+            return Mathf.Rad2Deg(Mathf.Atan(aspect * Mathf.Tan(Mathf.Deg2Rad(fovx) * (real_t)0.5)) * (real_t)2.0);
+        }
+
+        public real_t GetLodMultiplier()
+        {
+            if (IsOrthorgonal())
+            {
+                return GetViewportHalfExtends().x;
+            }
+            else
+            {
+                real_t zn = GetZNear();
+                real_t width = GetViewportHalfExtends().x * (real_t)2.0;
+                return (real_t)1.0 / (zn / width);
+            }
+        }
+
+        public int GetPixelsPerMeter(int forPixelWidth)
+        {
+            Vector3 result = Xform(new Vector3(1, 0, -1));
+
+            return (int)((result.x * (real_t)0.5 + (real_t)0.5) * forPixelWidth);
+        }
+
+        public Plane GetProjectionPlane(Planes plane)
+        {
+            Plane newPlane = plane switch
+            {
+                Planes.Near => new Plane(x.w + x.z, y.w + y.z, z.z + z.y, w.w + w.z),
+                Planes.Far => new Plane(x.w - x.z, y.w - y.z, z.z - z.y, w.w - w.z),
+                Planes.Left => new Plane(x.w + x.x, y.w + y.x, z.w + z.x, w.w + w.x),
+                Planes.Top => new Plane(x.w - x.y, y.w - y.y, z.w - z.y, w.w - w.y),
+                Planes.Right => new Plane(x.w - x.x, y.w - y.x, z.w - z.x, w.w - w.x),
+                Planes.Bottom => new Plane(x.w + x.y, y.w + y.y, z.w + z.y, w.w + w.y),
+            };
+            newPlane.Normal = -newPlane.Normal;
+            return newPlane.Normalized();
+        }
+
+        public Vector2 GetViewportHalfExtends()
+        {
+            var res = GetProjectionPlane(Planes.Near).Intersect3(GetProjectionPlane(Planes.Right), GetProjectionPlane(Planes.Top));
+            return new Vector2(res.Value.x, res.Value.y);
+        }
+
+        public real_t GetZFar()
+        {
+            Plane newPlane = new Plane(x.w + x.z, y.w + y.z, z.z + z.y, w.w + w.z);
```

Comment:
changed to `return GetProjectionPlane(Planes.Far).Normalized().D;`
and `return -GetProjectionPlane(Planes.Near).Normalized().D;`

Same result but different code, I would prefer this version to avoid copy-paste mistakes.

Reply:
If you are using `GetProjectionPlane` isn't `.Normalized` unnecessary? Since `GetProjectionPlane` already normalizes the result.

Reply:
yep